### PR TITLE
Extend chip-tool to send identify and identify-query commands

### LINKED ISSUE #2751: Missing support in chip-tool to test identify cluster
#### Problem
Missing support in chip-tool to test identify cluster

 #### Proposed Solution
implement it

### PR DESCRIPTION
**Pull request by bhaskar-apple**
Add two new classes for two new commands - identify and identify-query.
Also adds ability to handle cluster specific responses.
fixes #2751

### CODE REVIEW

File: examples/chip-tool/commands/common/ModelCommand.cpp [hunk 1]
```diff
@@ -109,20 +107,30 @@ void ModelCommand::ReceiveCommandResponse(ChipDeviceController * dc, PacketBuffe
...
+                 ChipLogError(chipTool, "Unexpected frame control byte: 0x%02x", frameControl));
     VerifyOrExit(sequenceNumber == 1, ChipLogError(chipTool, "Unexpected sequence number: %d", sequenceNumber));
 
-    switch (commandId)
+    isGlobalCommand = (frameControl == 8);
+    if (isGlobalCommand)
+    {
+        switch (commandId)
+        {
+        case 0x0b:
```
Comment (andy31415) [inline examples/chip-tool/commands/common/ModelCommand.cpp:120]: while we are touching this bit, could we add constants for the 0x0b and 0x01? They look so magical.

same for frameControl 8 and 9 above: we apparently needed comments to say what 8 and 9 mean.

Comment (bhaskar-apple) [reply examples/chip-tool/commands/common/ModelCommand.cpp:120]: Will do.

**Review by woody-apple — COMMENTED** [review]
approving, pending @andy31415 's suggestions

File: examples/chip-tool/commands/common/ModelCommand.cpp (hunk 1 above)
Comment (bhaskar-apple) [reply examples/chip-tool/commands/common/ModelCommand.cpp]: > approving, pending @andy31415 's suggestions

Made suggested changes.

File: examples/chip-tool/commands/common/ModelCommand.cpp [hunk 2]
```diff
@@ -89,6 +96,7 @@ void ModelCommand::ReceiveCommandResponse(ChipDeviceController * dc, PacketBuffe
     uint8_t frameControl;
     uint8_t sequenceNumber;
     uint8_t commandId;
+    bool isGlobalCommand = false;
 
     // Bit 3 of the frame control byte set means direction is server to client.
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/ModelCommand.cpp:101]: This comment shouldn't be here, and the comment about "Bit 3" a few lines down should move into `_isValidFrame` and be fixed to match the new reality.

File: examples/chip-tool/commands/common/ModelCommand.cpp [hunk 3]
```diff
@@ -81,6 +81,13 @@ void ModelCommand::SendCommand(ChipDeviceController * dc)
     dc->SendMessage(NULL, buffer);
 }
 
+bool _isValidFrame(uint8_t frameControl)
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/ModelCommand.cpp:84]: No leading `_`, please, but do put it in an anonymous namespace.

File: examples/chip-tool/commands/common/ModelCommand.cpp [hunk 4]
```diff
@@ -81,6 +81,13 @@ void ModelCommand::SendCommand(ChipDeviceController * dc)
...
+{
+#define kZCLGlobalCmdFrameControlHeader 8
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/ModelCommand.cpp:86]: These are going to be scoped outside the function anyway, so better to put them outside.

File: examples/chip-tool/commands/common/ModelCommand.cpp [hunk 5]
```diff
@@ -109,20 +117,28 @@ void ModelCommand::ReceiveCommandResponse(ChipDeviceController * dc, PacketBuffe
     commandId      = chip::Encoding::Read8(message);
     messageLen -= 3;
 
-    VerifyOrExit(frameControl == 8, ChipLogError(chipTool, "Unexpected frame control byte: 0x%02x", frameControl));
+    VerifyOrExit(_isValidFrame(frameControl), ChipLogError(chipTool, "Unexpected frame control byte: 0x%02x", frameControl));
     VerifyOrExit(sequenceNumber == 1, ChipLogError(chipTool, "Unexpected sequence number: %d", sequenceNumber));
 
-    switch (commandId)
+    isGlobalCommand = (frameControl == 8);
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/ModelCommand.cpp:123]: `kZCLGlobalCmdFrameControlHeader`

File: examples/chip-tool/commands/common/ModelCommand.h [hunk 6]
```diff
@@ -58,6 +63,7 @@ class ModelCommand : public NetworkCommand
     void ParseReadAttributeResponseCommandSuccess(uint16_t clusterId, uint16_t attrId, uint8_t * message,
                                                   uint16_t messageLen) const;
     void ParseReadAttributeResponseCommandFailure(uint16_t clusterId, uint16_t attrId, uint8_t status, uint16_t messageLen) const;
+    void ParseIdentifyClusterCommand(uint16_t endpointID, uint8_t * message, uint16_t messageLen) const;
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/ModelCommand.h:66]: Why is this here?  It's not used, right?

File: examples/chip-tool/commands/Identify-Cluster-Commands.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+class IdentifyQuery : public ModelCommand
+{
+public:
+    IdentifyQuery() : ModelCommand("identify-query"){};
+    size_t EncodeCommand(PacketBuffer * buffer, size_t bufferSize, uint16_t endPointId) override
+    {
+        return encodeIdentifyQueryCommand(buffer->Start(), bufferSize, endPointId);
+    }
+
+    void HandleClusterResponse(uint16_t clusterId, uint16_t endPointId, uint16_t commandId, uint8_t * message,
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/Identify-Cluster-Commands.h:33]: Shouldn't we check the `clusterId` somewhere here?

File: examples/chip-tool/commands/common/ModelCommand.cpp [hunk 8]
```diff
@@ -205,6 +221,12 @@ void ModelCommand::ParseReadAttributeResponseCommandFailure(uint16_t clusterId,
     return;
 }
 
+void ModelCommand::HandleClusterResponse(uint16_t clusterId, uint16_t endPointId, uint16_t commandId, uint8_t * message,
+                                         uint16_t messageLen) const
+{
+    ChipLogProgress(chipTool, "Received cmd %d for cluster %d", commandId, clusterId);
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/ModelCommand.cpp:227]: This should say something about it being unexpected to reach this, yes?

File: src/app/encoder.cpp [hunk 9]
```diff
@@ -173,4 +173,65 @@ uint16_t encodeReadOnOffCommand(uint8_t * buffer, uint16_t buf_length, uint8_t d
                                        1 /* attr_id_count */);
 }
 
+uint16_t _encodeIdentifyClusterCommand(uint8_t * buffer, uint16_t buf_length, uint8_t command_id, uint8_t destination_endpoint,
+                                       uint16_t * identify_duration)
+{
+    uint32_t result = 0;
+    // pick cluster id as 3 for now.
+    // pick source and destination end points as 1 for now.
```
Comment (bzbarsky-apple) [inline src/app/encoder.cpp:181]: We're not setting the destination endpoint to 1... it's passed in.  I guess `_encodeOnOffCommand` has the same bogus comment, sorry.

We should really share more code here with the on/off code.  Please file a followup for that?

File: src/app/chip-zcl-zpro-codec.h [hunk 10]
```diff
@@ -114,6 +114,27 @@ void printApsFrame(EmberApsFrame * frame);
  */
 uint16_t encodeApsFrame(uint8_t * buffer, uint16_t buf_length, EmberApsFrame * apsFrame);
 
+/**
+ * Idenfity cluster commands
```
Comment (mspang) [inline src/app/chip-zcl-zpro-codec.h:118]: Idenfity -> Identify

File: examples/chip-tool/commands/common/ModelCommand.cpp [hunk 11]
```diff
@@ -28,8 +28,8 @@
 using namespace ::chip;
 using namespace ::chip::DeviceController;
 
-#define CHECK_MESSAGE_LENGTH(rv)                                                                                                   \
-    VerifyOrExit(rv, ChipLogError(chipTool, "%s: Unexpected response length: %d", __FUNCTION__, messageLen));
+#define ZCL_DEFAULT_RESPONSE_COMMAND_ID 0x0b
```
Comment (andy31415) [inline examples/chip-tool/commands/common/ModelCommand.cpp:31]: nit: consider C++ 

```
namespace {
   constexpr uint8_t kDefaultResponseCommandId = 0x0b;
   constexpr uint8_t kReadAttributesResponseCommandId = 0x01;
}
```

or if these are enums we could have a CommandId enum.

#define does not feel as C++

File: examples/chip-tool/commands/common/ModelCommand.cpp (hunk 8 above)
Comment (BroderickCarlin) [inline examples/chip-tool/commands/common/ModelCommand.cpp:227]: As we look forward to fleshing out multi-endpoint support may want to also add that to the log:

```c++
ChipLogProgress(chipTool, "Received cmd %d for cluster %d on endpoint %d", commandId, clusterId, endPointId);
```

**Review by BroderickCarlin — APPROVED** [review]

File: src/app/encoder.cpp (hunk 9 above)
Comment (vivien-apple) [reply src/app/encoder.cpp:181]: I did a preemptive PR to fix that in #2766

File: examples/chip-tool/commands/common/ModelCommand.cpp (hunk 3 above)
Comment (vivien-apple) [reply examples/chip-tool/commands/common/ModelCommand.cpp:84]: Any reason to not use it a a private method of ModelCommand ?

File: examples/wifi-echo/server/esp32/main/gen/callback-stub.c [hunk 12]
```diff
@@ -2387,6 +2387,7 @@ void halSleepCallback(bool enter, SleepModes sleepMode) {}
  */
 bool emberAfPluginIdentifyStartFeedbackCallback(uint8_t endpoint, uint16_t identifyTime)
 {
+    emberAfPrintln(EMBER_AF_PRINT_IDENTIFY_CLUSTER, "Start idenitfy callback on endpoint %d time %d", endpoint, identifyTime);
```
Comment (vivien-apple) [inline examples/wifi-echo/server/esp32/main/gen/callback-stub.c:2390]: identify -> identify

File: examples/wifi-echo/server/esp32/main/gen/callback-stub.c [hunk 13]
```diff
@@ -2398,5 +2399,6 @@ bool emberAfPluginIdentifyStartFeedbackCallback(uint8_t endpoint, uint16_t ident
  */
 bool emberAfPluginIdentifyStopFeedbackCallback(uint8_t endpoint)
 {
+    emberAfPrintln(EMBER_AF_PRINT_IDENTIFY_CLUSTER, "Stop idenitfy callback on endpoint %d", endpoint);
```
Comment (vivien-apple) [inline examples/wifi-echo/server/esp32/main/gen/callback-stub.c:2402]: identify -> identify

File: examples/chip-tool/commands/common/ModelCommand.cpp (hunk 3 above)
Comment (bhaskar-apple) [reply examples/chip-tool/commands/common/ModelCommand.cpp:84]: I prefer anonymous namespace as well as this is mostly a utility method that is not necessarily integral to the class and its responsibilities.

File: examples/chip-tool/commands/common/ModelCommand.h (hunk 6 above)
Comment (bhaskar-apple) [reply examples/chip-tool/commands/common/ModelCommand.h:66]: Sorry this was a left over code from an earlier version of the PR that I hadn't posted.

**Review by rwalker-apple — APPROVED** [review]

**Comment by mspang** [comment]
Click Re-request review to avoid delays.

**Review by mspang — APPROVED** [review]
